File: algorithm-1/rewrite/consumer-choice.go
package main

import (
	"errors"
	"fmt"
	"math"
	"time"
)

/*
	this version should rewrite the intersection without the use of separate
	functions to find min and max, and should find all points

	additionally it is worth using the given bundle to eliminate the need to
	find both since we already have one of the two (under strict convexity/IDs)
*/

// define parameters and variables present in the optimization problem

var m, u float64
var t [2]float64
var p [2]float64

func B(x1, income float64) float64 {
	x2 := (income - (p[0] * x1)) / p[1]
	return x2
}

func I(x1, utility float64) float64 {
	x2 := math.Pow((utility / math.Pow(x1, t[0])), (1 / t[1]))
	return x2
}

// the rest of the functions are replacements for findIntersection

func difference(x1, income, utility float64) float64 {
	return I(x1, utility) - B(x1, income)
}

func derive(value, utility float64) float64 {
	h := 0.0001
	top := difference(value+h, m, utility) - difference(value, m, utility)
	bottom := h

	return top / bottom
}

func newtonsMethod(guess, utility float64) (float64, error) {
	xn := guess

	for n := 0; n <= 20; n++ {
		fxn := difference(xn, m, utility)
		Dfxn := derive(xn, utility)

		if math.Abs(fxn) < .001 {
			return xn, nil
		}

		if Dfxn == 0 {
			return -100, errors.New("newtonsMethod: zero division")
		}

		xn = xn - (fxn / Dfxn)
	}

	return -100, errors.New("newtonsMethod: exceeded max interations")
}

func adjust(lx, ux float64) float64 {

	x1 := (lx + ux) / 2
	x2 := B(x1, m)

	u := (math.Pow(x1, t[0])) * (math.Pow(x2, t[1]))

	return u
}

// close but nah, but still really close

func main() {
	// this is for the Marshallian case
	start := time.Now()

	m = 100

	t[0], t[1] = .3, .7
	p[0], p[1] = 3, 2

	u = math.Pow(.5*(m/p[0]), t[0]) * math.Pow(.5*(m/p[1]), t[1])

	var bounds [2]float64
	bounds[0], _ = newtonsMethod(.0001, u)
	bounds[1], _ = newtonsMethod(m/p[0], u)

	ctr := 0
	for bounds[1]-bounds[0] > .001 {
		u = adjust(bounds[0], bounds[1])
		bounds[0], _ = newtonsMethod(bounds[0], u)
		bounds[1], _ = newtonsMethod(bounds[1], u)
		fmt.Println(bounds)
		if ctr > 20 {
			break
		}
		ctr += 1
	}

	bundle := []float64{bounds[0], B(bounds[0], m)}
	elapsed := time.Since(start)

	for i := 0; i < 2; i++ {
		fmt.Printf("%.3f\t", bundle[i])
	}

	fmt.Printf("\ntime: %s", elapsed)
}

File: algorithm-1/rewrite/consumer_choice.jl
# rewrite in Julia because it is faster than Python and easier than Go

function marshallianDemand(m::Number, p::Array, t::Array)

    # define indifference
    function I(utl::Number,x1::Number)
        (utl*x1^(-t[1]))^(1/t[2])
    end

    # define affordability
    function B(x1::Number)
        (m-x1*p[1])/p[2]
    end

    # set initial affordable guess
    u = (.5*(m/p[1]))^(t[1])*(.5*(m/p[2]))^(t[2])

    function findIntersection(utl::Float64)
        function f(x1::Float64)
            I(utl,x1) - B(x1)
        end

        # derive using the limit definition
        function derive(x1::Float64)
            h = .0001
            return (f(x1+h)-f(x1))/h
        end

        function newtonsMethod(x0::Float64,maxiter::Integer=30)
            for _ in 1:maxiter
                yprime = derive(x0)
                if abs(yprime) < 1e-5
                    return x0
                end
                y = f(x0)
                x1 = x0 - y/yprime
                if abs(x1-x0) < 1e-5
                    return x1
                end
                x0 = x1
            end
            error("Max iteration exceeded")
        end
        return [newtonsMethod(.001),newtonsMethod(m/p[1])]
    end

    # adjust bounds
    function adjust(lb,ub)
        x1 = (lb+ub)/2
        x2 = B(x1)
        return x1^(t[1])*x2^(t[2])
    end

    bounds = findIntersection(u)

    # begin recursion
    while true
        if abs(bounds[2]-bounds[1]) > .0001
            u = adjust(bounds[1],bounds[2])
        else
            break
        end
        bounds = findIntersection(u)
    end
    return (bounds[1],I(u,bounds[1]))
end


function hicksianDemand(u::Number, p::Array, t::Array)

    # define indifference
    function I(x1::Number)
        (u*x1^(-t[1]))^(1/t[2])
    end

    # define affordability
    function B(inc::Float64,x1::Number)
        (inc-x1*p[1])/p[2]
    end

    # set initial indifferent guess
    x_i = [u^(1/(t[1]+t[2])),u^(1/(t[1]+t[2]))]
    m = sum([x_i[i]*p[i] for i in 1:2])

    function findIntersection(inc::Float64)
        function f(x1::Float64)
            I(x1) - B(inc,x1)
        end

        # derive using the limit definition
        function derive(x1::Float64)
            h = .0001
            return (f(x1+h)-f(x1))/h
        end

        function newtonsMethod(x0::Float64,maxiter::Integer=30)
            for _ in 1:maxiter
                yprime = derive(x0)
                if abs(yprime) < 1e-5
                    return x0
                end
                y = f(x0)
                x1 = x0 - y/yprime
                if abs(x1-x0) < 1e-5
                    return x1
                end
                x0 = x1
            end
            error("Max iteration exceeded")
        end
        return [newtonsMethod(.001),newtonsMethod(m/p[1])]
    end

    # adjust bounds
    function adjust(lb,ub)
        x1 = (lb+ub)/2
        x2 = I(x1)
        return x1*(p[1]) + x2*(p[2])
    end

    bounds = findIntersection(m)

    # begin recursion
    while true
        if abs(bounds[2]-bounds[1]) > .0001
            m = adjust(bounds[1],bounds[2])
        else
            break
        end
        bounds = findIntersection(m)
    end
    return (bounds[1],B(m,bounds[1]))
end

# benchmarking
@time marshallianDemand(100,[3 2],[.3 .7])
@time hicksianDemand(24.03519395349548,[3 2],[.3 .7])